engine/execution/provider/engine.go

```diff
@@ -24,150 +28,174 @@ import (
+	ready()
+	ticker := time.NewTicker(e.chdpRequestProcessInterval)
+	e.log.Debug().
+		Dur("process_inveral", e.chdpRequestProcessInterval).
+		Msg("process chunk data pack request worker started")
+
+	for {
+		select {
+		case <-ticker.C:
+			request, exists := e.chdpRequestsQueue.Pop()
+			if !exists {
+				e.log.Trace().Msg("worker did not find any chunk data pack request, sleeping till next cycle")
+				continue
+			}
+			lg := e.log.With().
+				Hex("chunk_id", logging.ID(request.ChunkId)).
+				Hex("origin_id", logging.ID(request.RequesterId)).Logger()
+			lg.Trace().Msg("chunk data pack request picked for processing")
+			e.onChunkDataRequest(request)
+			lg.Trace().Msg("chunk data pack request processed")
+		case <-ctx.Done():
+			e.log.Trace().Msg("processing chunk data pack request worker terminated")
+			return
+		}
+	}
 }
 
 func (e *Engine) Process(channel channels.Channel, originID flow.Identifier, event interface{}) error {
-	return e.unit.Do(func() error {
-		return e.process(originID, event)
-	})
+	return e.process(originID, event)
 }
 
 func (e *Engine) process(originID flow.Identifier, event interface{}) error {
-	ctx := context.Background()
 	switch v := event.(type) {
 	case *messages.ChunkDataRequest:
-		e.onChunkDataRequest(ctx, originID, v)
+		pushed := e.chdpRequestsQueue.Push(v.ChunkID, originID)
+		e.log.Info().
+			Hex("chunk_id", logging.ID(v.ChunkID)).
+			Hex("origin_id", logging.ID(originID)).
+			Bool("enqueued", pushed).
+			Msg("chunk data pack request received")
 	default:
 		return fmt.Errorf("invalid event type (%T)", event)
 	}
-
 	return nil
 }
 
-// onChunkDataRequest receives a request for the chunk data pack associated with chunkID from the
-// requester `originID`. If such a chunk data pack is available in the execution state, it is sent to the
-// requester.
-func (e *Engine) onChunkDataRequest(
-	ctx context.Context,
-	originID flow.Identifier,
-	req *messages.ChunkDataRequest,
-) {
-
+// onChunkDataRequest receives a request for a chunk data pack,
+// and if such a chunk data pack is available in the execution state, it is sent to the requester node.
+func (e *Engine) onChunkDataRequest(request *mempool.ChunkDataPackRequest) {
 	processStart := time.Now()
 
-	// extracts list of verifier nodes id
-	chunkID := req.ChunkID
-
 	lg := e.log.With().
-		Hex("origin_id", logging.ID(originID)).
-		Hex("chunk_id", logging.ID(chunkID)).
+		Hex("origin_id", logging.ID(request.RequesterId)).
+		Hex("chunk_id", logging.ID(request.ChunkId)).
 		Logger()
-
 	lg.Info().Msg("received chunk data pack request")
 
 	// increases collector metric
 	e.metrics.ChunkDataPackRequested()
+	chunkDataPack, err := e.execState.ChunkDataPackByChunkID(request.ChunkId)
 
-	chunkDataPack, err := e.execState.ChunkDataPackByChunkID(ctx, chunkID)
 	// we might be behind when we don't have the requested chunk.
 	// if this happen, log it and return nil
 	if errors.Is(err, storage.ErrNotFound) {
```

Comment:
Refactored in https://github.com/onflow/flow-go/pull/2951/commits/9da2a9840aa14a56ba7776fff64224cd712f0a40. However, not pretty comfortable with this change, as one we go permissionless with VNs, it allows a malicious VN to conduct a DoS attack by sending invalid `OriginId`s.

Reply:
> one we go permissionless with VNs, it allows a malicious VN to conduct a DoS attack by sending invalid OriginIds

I agree - If we don't also differentiate between expected and unexpected error conditions, then DOS attacks are made easier by this suggestion.

From what I can see, the unexpected errors we want to propagate are:
* storage error reading chunk https://github.com/onflow/flow-go/commit/9da2a9840aa14a56ba7776fff64224cd712f0a40#diff-99567a576e6710d390c77cd207573be724e407c3fab2d99f3cbd4bc6334e7572R257
  * this is already handled correctly, and isn't subject to DOS attacks
* storage error checking if requester is authorized https://github.com/onflow/flow-go/commit/9da2a9840aa14a56ba7776fff64224cd712f0a40#diff-99567a576e6710d390c77cd207573be724e407c3fab2d99f3cbd4bc6334e7572R262
  * this could cause a crash for requests with either a non-existent or unknown chunk ID, or unknown Origin IDs
  * to fix this, we should document all expected error return types from `ensureAuthorized`, handle the expected cases within `onChunkDataRequest`, and bubble up the unexpected errors (eg. storage error besides `ErrNotFound`)

Reply:
Per our conversation, created this https://github.com/dapperlabs/flow-go/issues/6363 to address proper error handing on the provider engine separately. Also, for now, reverted back error handling to the same state as `master` hence not causing crash incidents.